accepted/future-releases/unified-collections/feature-specification.md

```diff
@@ -0,0 +1,802 @@
+    constant elements, or at least potentially constant elements.
+
+*   An `expressionElement` is a constant element if its expression is a constant
+    expression, and a potentially constant element if it's expression is a
+    potentially constant expression.
+
+*   A `mapEntry` is a constant element if both its key and value expression are
+    constant expressions, and a potentially constant element if both are
+    potentially constant expressions.
+
+*   A `spreadElement` is a constant element if its expression is constant and it
+    evaluates to a constant `List`, `Set` or `Map` instance originally created
+    by a list, set or map literal. It is a potentially constant element if the
+    expression is potentially constant expression.
+
+*   An `ifElement` is constant if its condition is a constant expression
+    evaluating to a Boolean value and either:
+
+    *   If the condition evaluates to `true`, the then element is a constant
+        expression, and any else element is a potentially constant element.
+
+    *   If the condition evaluates to `false`, then the then element is a
+        potentially constant element and any else element is a constant element.
+
+*   A `forElement` is never a constant element. `for` cannot be used in constant
+    set or map literals.
+
+*   It is a compile-time error to have duplicate values in a set literal or
+    duplicate keys in a map literal, according to behavior specified in the
+    runtime semantics.
+
+*   When evaluating a const collection literal, if there is an existing const
+    list, map, or set literal that contains the same series of values or entries
+    in the same order, then that object is used instead of producing a new one.
+    *In other words, constants are canonicalized.* Note that maps are only
+    canonicalized if they contain the same keys *in the same order*.
+
+### Constant equality safe
+
+A value is constant equality safe iff:
+
+*   It is an instance of a class that does not override `Object.operator==`, or
+
+*   it is an instance of `int` or `String`, or
+
+*   it is an instance of `Symbol` originally created by a symbol literal or a
+    constant invocation of the `Symbol` constructor, or
+
+*   it is an instance of `Type` originally created by evaluating a type literal
+    expression (an identifier or qualified identifier denoting a class, mixin or
+    type alias declaration, that is evaluated as an expression).
+
+All constant equality safe values are also valid constant values.
+
+## Dynamic Semantics
+
+Spread, `if`, and `for` behave similarly across all collection types. To
+simplify the spec, there is a single procedure used for all kinds of
+collections. This recursive procedure builds up either a `result` sequence of
+either values or key-value pair map entries. Then a final step handles those
+appropropriately for the given collection type.
+
+### To evaluate a collection `element`:
```

Comment:
Here we need to say that evaluation of an `element` can complete either with a thrown error, or successfully.
(Like expressions can evaluate to a thrown error or a with a value).
Something like:

> Evaluation of an `element` either throws an error and a stack trace, or it completes successfully.
> If evaluation of an element includes evaluation of an expression, and that expression evaluation throws an error and stack trace, then, unless otherwise stated, evaluation of the element ends there and throws the same error and stack trace.
> Otherwise, if nothing else is stated, evaluation of an element is successful when the algorithm reaches the end.

Reply:
Does this really need to be spelled out? I don't understand what this adds.

Reply:
It'll be needed in the specification, because we are introducing a new syntactic category that we haven't said anything about yet. It's probably not needed here.